packages/editor/src/lib/utils/reorderShapes.ts

```diff
@@ -147,14 +149,48 @@ function reorderToFront(moving: Set<TLShape>, children: TLShape[], changes: TLSh
 	}
 }
 
+function getVerticesInPageSpace(editor: Editor, shape: TLShape) {
+	const geo = editor.getShapeGeometry(shape)
+	const pageTransform = editor.getShapePageTransform(shape)
+	if (!geo || !pageTransform) return null
+	return pageTransform.applyToPoints(geo.vertices)
+}
+
+function filterOverlappingShapes(editor: Editor, moving: Set<TLShape>, children: TLShape[]) {
+	const movingVertices = Array.from(moving)
+		.map((shape) => {
+			const vertices = getVerticesInPageSpace(editor, shape)
+			if (!vertices) return null
+			return { shape, vertices }
+		})
+		.filter(Boolean) as { shape: TLShape; vertices: Vec[] }[]
+
+	// only keep overlapping shapes
+	return children.filter((child) => {
+		const vertices = getVerticesInPageSpace(editor, child)
+		if (!vertices) return false
+		return movingVertices.some((other) => {
+			return polygonsIntersect(other.vertices, vertices)
+		})
+	})
+}
+
 /**
  * Reorders the moving shapes forward in the parent's children.
  *
+ * @param editor The editor
  * @param moving The set of shapes that are moving
  * @param children The parent's children
  * @param changes The changes array to push changes to
  */
-function reorderForward(moving: Set<TLShape>, children: TLShape[], changes: TLShapePartial[]) {
+function reorderForward(
+	editor: Editor,
+	moving: Set<TLShape>,
+	children: TLShape[],
+	changes: TLShapePartial[]
+) {
+	children = filterOverlappingShapes(editor, moving, children)
```

Comment:
Possibly handled by the jitter that we add to fractional indexes but I don't know if I like relying on that. Here's a scenario:
1. There are 3 shapes, A, B, C, & D. They have indexes 1, 2, 3, & 4 respectively. A+B+D overlap, but C does not.
2. The user tries to bring A forward. The children array is `[A, B, D]`.
3. We find B as the first selecting shape, then D as its next item in the array
4. `getIndexBetween(2, 4) => 3`
5. A now has the same index as C

Instead of filtering out all the children at the start, should we apply the filtering in the skipping/selecting loop? That way, when we look up the next child, we'll automatically get the correct one and avoid any chance of repeated IDs.

Reply:
good suggestion I'll give that a go